### PR TITLE
Replace PRNG with additional RO function

### LINKED ISSUE #36: Replace PRNG with RO for batched proof
As Hugo notes:
```
 2. d1,...dn <- PRNG(seed,salt,info,n)

> PRNGs inputs are just a seed, usually a secret one. There are no guarantees
> about the pseudorandomness of the output for a public seed. Additional items
> such as salt, info, n, are not accommodated at all in a PRNG. You could put
> these under H_4 or use a (variable-length output) PRF instead of PRNG.
> But also in the PRF case, non-secret key does not guarantee pseudorandomness.
> What you seem to need is a RO which on inputs G,Y,[Mi,Zi],salt,info,n outputs 
> d1,...dn (is there a proof that batching is secure when both parties derive d_i
> using a RO?)

> Btw, independently of the above issues, if you let the prover choose the salt,
> this can help him (at least in theory) to cheat - i.e., finding a set of
> "misleading"  d's.
```
I think it will be easiest to introduce an additional RO `H_4` that handles the generation of `d1,...,d4` here.

We would also need to clarify the best instantiations for doing this, it may be that we stick with SHAKE-SHA256 and HKDF-Expand-SHA256 (this should also be clarified as we currently just say HKDF-256). The notion of the `salt` should be replaced with a domain separation label.

### PR DESCRIPTION
**Pull request by alxdavids**
Fixes #36
- Usage of PRNG is unnecessary.
- We now instantiate with a generic RO
- Specific ciphersuites are unchanged
- HKDF -> HKDF_Expand

### CODE REVIEW

File: draft-sullivan-cfrg-oprf.md [hunk 1]
```diff
@@ -923,22 +922,28 @@ Steps:
...
-multiple blinded PRF inputs [Mi] for i in 1...n. Then P computes the array [Zi]
-and replaces DLEQ_Generate with Batched_DLEQ_Generate over these arrays. The
+multiple blinded PRF inputs [ Mi ] for i in 1...n. Then P computes the array [Zi
+] and replaces DLEQ_Generate with Batched_DLEQ_Generate over these arrays. The
 same applies to the algorithm VOPRF_Sign. The same applies for replacing
 DLEQ_Verify with Batched_DLEQ_Verify when V verifies the response from P and
 during the algorithm VOPRF_Verify.
 
-## PRNG and resampling
+## Random oracle instantions for proofs
```
Comment (chris-wood) [inline draft-sullivan-cfrg-oprf.md]:
```suggestion
## Random oracle instantiations for proofs
```

File: draft-sullivan-cfrg-oprf.md [hunk 2]
```diff
@@ -923,22 +922,28 @@ Steps:
...
 
-Any function that satisfies the security properties of a pseudorandom number
-generator can be used for computing the batched DLEQ proof. For example,
-SHAKE-256 {{SHAKE}} or HKDF-SHA256 {{RFC5869}} would be reasonable choices for
-groups that have an order of 256 bits.
+We can instantiate the random oracle function H_4 using the same hash function
+that is used for H_1,H_2,H_3. For H_5, we can also use a similar instantiation,
+or we can use a stream-based output generator. For example, for groups with an
```
Comment (chris-wood) [inline draft-sullivan-cfrg-oprf.md]:
```suggestion
or we can use a variable-length output generator. For example, for groups with an
```

File: draft-sullivan-cfrg-oprf.md [hunk 3]
```diff
@@ -923,22 +922,28 @@ Steps:
...
+order of 256-bit, valid instantiations include functions such as SHAKE-256
+{{SHAKE}} or HKDF_Expand-SHA256 {{RFC5869}}.
```
Comment (chris-wood) [inline draft-sullivan-cfrg-oprf.md]: Is one preferred over the other? Why? (We should probably include some rationale text here.)

File: draft-sullivan-cfrg-oprf.md [hunk 4]
```diff
@@ -923,22 +922,28 @@ Steps:
...
 
-We note that the PRNG outputs d1,...,dn must be smaller than the order of the
-group/curve that is being used. Resampling can be achieved by increasing the
-value of the iterator that is used in the info field of the PRNG input.
+In addition if a function with larger output than the order of the base field is
+used, we note that the outputs of H_5 (d1,...,dn) must be smaller than this
+order. If any di that is sampled is larger than then order, then we should
+resample until a di' is sampled that is valid.
+
+In these cases, the iterating integer i is increased to i' until such di' is
```
Comment (chris-wood) [inline draft-sullivan-cfrg-oprf.md]:
```suggestion
In these cases, the iterating integer i is increased monotonically to i' until such di' is
```

Comment (chris-wood) [inline draft-sullivan-cfrg-oprf.md]: In general, this is underspecified. We should leave a TODO here at least to write this out.

File: draft-sullivan-cfrg-oprf.md [hunk 5]
```diff
@@ -958,7 +963,7 @@ ciphersuites demonstrating 128 bits of security.
 - H_2: SHA256
 - H_3: SHA256
 - H_4: SHA256
-- PRNG: HKDF-SHA256
+- H_5: HKDF_Expand-SHA256
```
Comment (chris-wood) [inline draft-sullivan-cfrg-oprf.md]: It's typically written as `HKDF-Expand`, not `HKDF_Expand`. Can we use that instead?

File: draft-sullivan-cfrg-oprf.md [hunk 6]
```diff
@@ -1120,9 +1125,8 @@ Z_0: 043ab5ccb690d844dcb780b2d9e59126d62bc853ba01b2c339ba1c1b78c03e4b6adc5402f7\
 Z_1: 04647e1ab7946b10c1c1c92dd333e2fc9e93e85fdef5939bf2f376ae859248513e0cd91115\
```
Comment (chris-wood) [inline draft-sullivan-cfrg-oprf.md]: Are these test vectors correct with HKDF? We should probably remove them, temporarily, if not.

File: draft-sullivan-cfrg-oprf.md [hunk 7]
```diff
@@ -923,22 +922,28 @@ Steps:
 ## Modified protocol execution
 
 The VOPRF protocol from Section {{protocol}} changes to allow specifying
-multiple blinded PRF inputs [Mi] for i in 1...n. Then P computes the array [Zi]
-and replaces DLEQ_Generate with Batched_DLEQ_Generate over these arrays. The
+multiple blinded PRF inputs [ Mi ] for i in 1...n. Then P computes the array [Zi
```
Comment (chris-wood) [inline draft-sullivan-cfrg-oprf.md]:
```suggestion
multiple blinded PRF inputs [ Mi ] for i in 1...n. Then P computes the array [Zi]
```

File: draft-sullivan-cfrg-oprf.md [hunk 8]
```diff
@@ -923,22 +922,28 @@ Steps:
...
+] and replaces DLEQ_Generate with Batched_DLEQ_Generate over these arrays. The
```
Comment (chris-wood) [inline draft-sullivan-cfrg-oprf.md]:
```suggestion
and replaces DLEQ_Generate with Batched_DLEQ_Generate over these arrays. The
```

File: draft-sullivan-cfrg-oprf.md (hunk 3 above)
Comment (alxdavids) [reply draft-sullivan-cfrg-oprf.md]: I don't think either is preferred, they're just examples of functions that you might want to use. We give ciphersuites below that use HKDF_Expand, is that enough?

File: draft-sullivan-cfrg-oprf.md (hunk 4 above)
Comment (alxdavids) [reply draft-sullivan-cfrg-oprf.md]: Good point, I will add this.

File: draft-sullivan-cfrg-oprf.md (hunk 5 above)
Comment (alxdavids) [reply draft-sullivan-cfrg-oprf.md]: Sure

File: draft-sullivan-cfrg-oprf.md (hunk 6 above)
Comment (alxdavids) [reply draft-sullivan-cfrg-oprf.md]: Yeah they are correct with HKDF (rather than Shake)

**Comment by chris-wood** [comment]
Thanks for the changes! Once the conflicts are resolved, feel free to merge.

**Pull request merged**